src/codechicken/nei/recipe/GuiRecipe.java

```diff
@@ -193,20 +195,140 @@ private void checkYShift() {
         yShift = handlerInfo == null ? 0 : handlerInfo.getYShift();
     }
 
-
     public void setRecipePage(int idx) {
+        setRecipePage(idx, 0);
+    }
+
+    public void setRecipePage(int idx, int idp) {
         recipetype = idx;
         if (recipetype < 0) recipetype = currenthandlers.size() - 1;
         else if (recipetype >= currenthandlers.size()) recipetype = 0;
 
         handler = currenthandlers.get(recipetype);
         handlerInfo = getHandlerInfo(handler);
-        page = 0;
+        page = Math.min(idp, (handler.numRecipes() - 1) / getRecipesPerPage());
         recipeTabs.calcPageNumber();
         checkYShift();
         initOverlayButtons();
     }
 
+    public void openTargetRecipe(String recipeId) 
+    {
+
+        if (recipeId == "") {
+            return;
+        }
+
+        String[] expected = recipeId.split("\\|", 2);
+        Integer recipetype = 0;
+        Integer page = 0;
+
+        for (int j = 0; j < currenthandlers.size(); j++) {
+            IRecipeHandler localHandler = currenthandlers.get(j);
+            HandlerInfo localHandlerInfo = getHandlerInfo(localHandler);
+
+            if (localHandlerInfo.getHandlerName().equals(expected[0])) {
+                recipetype = j;
+
+                for (int i = 0; i < localHandler.numRecipes(); i++) {
+                    if (createRecipeId(localHandler, i).equals(expected[1])) {
```

Comment:
This seems... very heavy handed on the allocation side to find the right page. We're creating a string to compare, and then throwing it away.  Is there another ID we can use that already exists, or some way to cache this?  Some of these handlers can have a large number of items and this has the potential to re-introduce slowness back into NEI.

Reply:
I tried to compare nbt, but the code became too complicated to understand. Also more strings will be created on handlers with more variants for recipes for one item. And this is not so many handlers